example/bench/bench.cpp

```diff
@@ -33,16 +35,26 @@ int main(int argc, char** argv) {
     bool is_root = true;
 
     try {
+        arb::proc_allocation resources;
+        if (auto nt = sup::get_env_num_threads()) {
+            resources.num_threads = *nt;
+        }
+        else {
+            resources.num_threads = sup::thread_concurrency();
+        }
+
 #ifdef ARB_MPI_ENABLED
```

Comment:
This thread count code is repeated in each example, and will likely appear in pretty much every C++ code that uses the library. It's boilerplate code that could be wrapped up with e.g. `sup::get_default_threads()` or similar; people can still use the more fundamental functions if they wish.

Reply:
It is repeated in each example, but I don't have a problem with this because:
* I think this makes it much clearer what is going on: test the environment variable, then attempt to inspect the CPU set if the environment variable isn't set.
* the examples are meant to be stand alone, so a bit of repeated code isn't a crime.

I really feel that combining this logic into a single function obscures what is going on too much.
This level of granularity feels "about right" for what the `sup` library is trying to achieve: small well-defined tools/types/functions that are useful for all examples, tests, etc.

Reply:
I think though that almost every such example will have similar code, and a bunch of user code will too. By providing an easy-to-use default function, we make it easy for library users also to do the reasonable thing by default.

`sup` contains mainly things that we want to use in common across the examples and tests, which should not form part of the core interface. Wrapping boilerplate code is part of its very purpose — I really don't see this as a problem at all.

This thread counting code and the GPU negotiation code will be in whatever we will call our installed-with-Arbor helper library; any obscurity in its functionality can be well addressed simply with documentation, IMO.

Reply:
I must admit I am a bit uneasy suggesting that `ARB_NUM_THREADS` is a reasonable environment variable for users ;)
More seriously, user code might be similar, but would in all likelihood use different environment variables.
The examples might share code via `sup`, but at the same time they should be as simple to understand stand-alone. Because nobody reads the docs (they just create a ticket).

This feels a bit like 'half of one, six of the other'. I can change it, but I think that both of our positions are fairly reasonable.